### PR TITLE
External lb

### PR DESCRIPTION
**Pull request by sgaoshang**
Add workflow to enable vSphere external lb

### CODE REVIEW

**Comment by sgaoshang** [comment]
pj-rehearse

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+      echo "    server $ip $ip:443 check"
+    done
+    )
+EOF
+
+# scp new_pull_secret credential to bastion host
+scp -o UserKnownHostsFile=/dev/null -o IdentityFile="${SSH_PRIV_KEY_PATH}" -o StrictHostKeyChecking=no "${haproxy_cfg}" ${BASTION_SSH_USER}@${BASTION_IP}:${bastion_haproxy_cfg}
+
+# Create /etc/haproxy/haproxy.cfg and load it by restart haproxy.service in bastion host
+ssh -o UserKnownHostsFile=/dev/null -o IdentityFile="${SSH_PRIV_KEY_PATH}" -o StrictHostKeyChecking=no ${BASTION_SSH_USER}@${BASTION_IP} "sudo systemctl restart haproxy.service"
```
Comment (jinyunma) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh:102]: seems missing step to update /etc/haproxy/haproxy.cfg after copying it to /tmp folder on bastion host.

Comment (sgaoshang) [reply ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh:102]: @jinyunma restart haproxy.service will reload /tmp/haproxy.cfg, I need to correct the comment, thanks for the catch.

File: ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/external-lb/provision/cucushift-installer-rehearse-vsphere-ipi-external-lb-provision-chain.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+chain:
+  as: cucushift-installer-rehearse-vsphere-ipi-external-lb-provision
+  steps:
+  - chain: ipi-conf-vsphere
+  - ref: ignition-bastionhost
+  - ref: ignition-bastionhost-external-lb
```
Comment (jinyunma) [inline ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/external-lb/provision/cucushift-installer-rehearse-vsphere-ipi-external-lb-provision-chain.yaml:6]: How about renaming to `ignition-bastionhost-haproxy`?

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+          "DNSName": "'${nlb_dnsname}'",
+          "EvaluateTargetHealth": false
+          }'
+    apps_dns_target=$api_dns_target
+  fi 
+
+else
+  # Configure DNS direct to respective VIP
+  api_dns_target='"TTL": 60,
+        "ResourceRecords": [{"Value": "'${vips[0]}'"}]'
```
Comment (jinyunma) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh:185]: here should change to bastion host ip for api dns.

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  apps_dns_target='"TTL": 60,
+        "ResourceRecords": [{"Value": "'${vips[1]}'"}]'
```
Comment (jinyunma) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh:187]: here should change to bastion host ip for ingress dns.

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+mapfile -t vips < "${SHARED_DIR}"/vips.txt
+
+hosted_zone_id="$(aws route53 list-hosted-zones-by-name \
+            --dns-name "${base_domain}" \
+            --query "HostedZones[? Config.PrivateZone != \`true\` && Name == \`${base_domain}.\`].Id" \
+            --output text)"
+echo "${hosted_zone_id}" > "${SHARED_DIR}/hosted-zone.txt"
+
+if [ "${JOB_NAME_SAFE}" = "launch" ]; then
+  # If this is a `launch` and we are in the multi-zone range, we also
```
Comment (jinyunma) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh:151]: For ipi + external lb workflow, do we need `if` block here?

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+      $apps_dns_target
+      }
+}]}
+EOF
+
+# api-int record is needed for Windows nodes
+# TODO: Remove the api-int entry in future
+echo "Creating batch file to destroy DNS records"
+
+cat > "${SHARED_DIR}"/dns-delete.json <<EOF
```
Comment (jinyunma) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh:225]: Here just only update existing api/ingress dns records, I guess they will be deleted when calling step [ipi-deprovision-vsphere-dns](https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/deprovision/vsphere/dns/ipi-deprovision-vsphere-dns-commands.sh#L33). So do we need to overwrite file `"${SHARED_DIR}"/dns-delete.json` here?

File: ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/external-lb/deprovision/cucushift-installer-rehearse-vsphere-ipi-external-lb-deprovision-chain.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+chain:
+  as: cucushift-installer-rehearse-vsphere-ipi-external-lb-deprovision
+  steps:
+  - ref: gather-proxy
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/vsphere/ipi/external-lb/deprovision/cucushift-installer-rehearse-vsphere-ipi-external-lb-deprovision-chain.yaml:4]: In this workflow, did not involve proxy, so sounds we do not need this step.

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-ref.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
+ref:
+  as: ipi-conf-external-lb
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-ref.yaml:2]: Seem like this step is doing a post-action to configure an external lb for a cluster, but not configure install-config, so suggest to rename it to `vsphere-external-lb`

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-ref.metadata.json [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{
+	"path": "ipi/conf/external-lb/ipi-conf-external-lb-ref.yaml",
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-ref.metadata.json:2]: Seem like this step is doing a post-action to configure an external lb for a cluster, but not configure install-config, so suggest to rename it to `vsphere/external-lb`

File: ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    "Action": "UPSERT",
+    "ResourceRecordSet": {
+      "Name": "*.apps.$cluster_domain.",
+      "Type": "A",
+      "TTL": 60,
+      "ResourceRecords": [{"Value": "${BASTION_IP}"}]
+      }
+}]}
+EOF
+id=$(aws route53 change-resource-record-sets --hosted-zone-id "$hosted_zone_id" --change-batch file:///"${SHARED_DIR}"/dns-update.json --query '"ChangeInfo"."Id"' --output text)
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/external-lb/ipi-conf-external-lb-commands.sh:132]: I did not see the step of feeding aws credentails to run the aws command in this script.

**Comment by jianlinliu** [comment]
@sgaoshang @jinyunma for this case, shall we pass the external lb dns to e2e testing instead of the original one in kubeconfig file, right?

**Comment by jinyunma** [comment]
> shall we pass the external lb dns to e2e testing instead of the original one in kubeconfig file, right?
nope, just replace api vip and ingress vip to bastion host ip for api/ingress dns, then it will access cluster through external lb on ports 6443, 443, and 80 which we defined in haproxy.cfg.

**Comment by jianlinliu** [comment]
Okay, got it on `Updating DNS records` lines. Thanks.

**Pull request merged**